dhcpv4/server.go

```diff
@@ -90,49 +97,52 @@ func (s *Server) ActivateAndServe() error {
 		s.conn = conn
 	}
 	s.connMutex.Unlock()
-	defer s.Close()
 	var (
 		pc *net.UDPConn
 		ok bool
 	)
 	if pc, ok = s.conn.(*net.UDPConn); !ok {
-		return fmt.Errorf("Error: not an UDPConn")
+		return fmt.Errorf("error: not an UDPConn")
 	}
 	if pc == nil {
-		return fmt.Errorf("ActivateAndServe: Invalid nil PacketConn")
+		return fmt.Errorf("ActivateAndServe: invalid nil PacketConn")
 	}
 	log.Printf("Server listening on %s", pc.LocalAddr())
 	log.Print("Ready to handle requests")
-	for {
-		select {
-		case <-s.shouldStop:
-			break
-		case <-time.After(time.Millisecond):
-		}
-		pc.SetReadDeadline(time.Now().Add(time.Second))
-		rbuf := make([]byte, 4096) // FIXME this is bad
-		n, peer, err := pc.ReadFrom(rbuf)
-		if err != nil {
-			switch err.(type) {
-			case net.Error:
-				if err.(net.Error).Timeout() {
-					return err
+	go func() {
```

Comment:
Do we need a goroutine here? We are already using 1 goroutine per response handler.

Reply:
...I wanted to say a channel, sorry. Fixing

Reply:
discussed offline - as you suggest it's not necessary to make the function non-blocking here. It can be easily achieved by the caller, and error handling is easier.
I have reverted the goroutine part and left only the bug fixes and code consolidation